### PR TITLE
Default HTTP status code when not set

### PR DESCRIPTION
**Pull request by 82marbag**
`RestJsonHttpResponseCodeDefaultsToModeledCode` now succeeds.

See https://github.com/awslabs/smithy/issues/1098

## Testing
`./gradlew -P modules='rest_json' :codegen-server-test:test`

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by LukeMathWalker — APPROVED** [review]

**Review by crisidev — APPROVED** [review]

**Review by jjant — APPROVED** [review]

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt [hunk 1]
```diff
@@ -678,7 +683,7 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
             if (symbolProvider.toSymbol(binding.member).isOptional()) {
                 rustTemplate(
                     """
-                    .ok_or(#{ResponseRejection}::MissingHttpStatusCode)?
+                    .unwrap_or($fallbackStatusCode)
```
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:686]: We can now remove `MissingHttpStatusCode` from `rust-runtime/aws-smithy-http-server/src/rejection.rs`.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt [hunk 2]
```diff
@@ -538,15 +539,18 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
         Attribute.AllowUnusedMut.render(this)
         rustTemplate("let mut builder = #{http}::Response::builder();", *codegenScope)
         serverRenderResponseHeaders(operationShape)
+        // Fallback to the default code of `@http`, 200.
+        val httpTraitDefaultStatusCode = HttpTrait
```
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:543]: Looking at [`HttpTrait.java`](https://github.com/awslabs/smithy/blob/f5a0893cd0f9983d0bd5697db7c226c1d379ef92/smithy-model/src/main/java/software/amazon/smithy/model/traits/HttpTrait.java#L119), it looks like `operationShape.getTrait<HttpTrait>()?.code` should be 200 if no code was modeled, so we shouldn't have to be building an artificial `HttpTrait` here.

Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:543]: That's the (private) builder with only a setter (no getter) for the status code. We still have to build an artificial HttpTrait

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt (hunk 1 above)
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:686]: https://github.com/awslabs/smithy-rs/issues/2229

**Pull request merged**